bridge-worker/server.go

```diff
@@ -0,0 +1,253 @@
+) *Server {
+	s := &Server{
+		logger:               logger,
+		chainID:              chainID,
+		mezoBridge:           mezoBridge,
+		store:                store,
+		assetsUnlockEndpoint: assetsUnlockedClient,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("POST /attestations", s.submitAttestation)
+
+	s.server = &http.Server{
+		Addr:              fmt.Sprintf(":%d", port),
+		Handler:           mux,
+		ReadHeaderTimeout: time.Second,
+	}
+
+	return s
+}
+
+func (s *Server) Start() {
+	s.logger.Info("http server starting", "address", s.server.Addr)
+	go func() {
+		if err := s.server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			// should we panic here? likely yes
+			s.logger.Error("server failed to start", "error", err)
+		}
+	}()
+}
+
+func (s *Server) Stop(ctx context.Context) error {
+	return s.server.Shutdown(ctx)
+}
+
+func (s *Server) submitAttestation(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	req, err := readSubmitAttestationRequest(r)
+	if err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if err := req.Validate(); err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	events, err := s.assetsUnlockEndpoint.GetAssetsUnlockedEvents(r.Context(), req.Entry.UnlockSequence, req.Entry.UnlockSequence.AddRaw(1))
+	if err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// should have exactly one event
+	if len(events) != 1 {
+		s.writeSubmitAttestationError(w, errors.New("invalid attestation"), http.StatusBadRequest)
+		return
+	}
+
+	if !events[0].Equal(req.Entry) {
+		s.writeSubmitAttestationError(w, errors.New("invalid attestation"), http.StatusBadRequest)
+		return
+	}
+
+	// first recover the address out of the signature
+	address, err := s.recoverAddress(req.Entry, req.Signature)
+	if err != nil {
+		s.writeSubmitAttestationError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	// now validate the address is a registered validator address
+	index, err := s.mezoBridge.BridgeValidatorIDs(address)
+	if err != nil {
+		s.logger.Error("couldn't get bridge validator ID", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// if default value, then address is not a validator
+	if index == 0 {
+		s.writeSubmitAttestationError(w, errors.New("not an authorized validator"), http.StatusUnauthorized)
+		return
+	}
+
+	// then the attestation has not been confirmed yet
+	isConfirmed, err := s.mezoBridge.ConfirmedUnlocks(req.Entry.UnlockSequence.BigInt())
+	if err != nil {
+		s.logger.Error("couldn't confirm unlock", "error", err)
+		s.writeSubmitAttestationError(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	// if already confirmed, nothing to do
+	if isConfirmed {
+		s.writeSubmitAttestationError(w, errors.New("already confirmed"), http.StatusBadRequest)
```

Comment:
This is not an error. Can happen from time to time if there are time drifts and a honest validator submits its signature with a delay. I would return a 200 in this case.